### DISCUSSION TITLE
How to delete all small images inserted on each page?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#3313] [General]
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/6a637fc6-4e19-4ab9-8c0b-0420960acf67)

**Comment by JorjMcKie** [comment]
If these are true images (with an xref etc.) you can delete them via `page.delete_image(xref)`.
This actually is a replacement with a super-small transparent pixmap.

Otherwise you can use redactions, which now can also erase vector graphics (v1.24.0).

**Comment by nissansz** [comment]
Can you share a full code to remove for example above file?

**Reply by JorjMcKie** [reply]
```python
for page in doc:
    for i in page.get_images():
        page.delete_image(i[0])
```

**Comment by nissansz** [comment]
thank you, how to add a condition to exclude some images, like speicified width, height etc.

**Reply by JorjMcKie** [reply]
Look at the documentation of page.get_images() output, i[0] is just the xref of the whole item.

**Comment by nissansz** [comment]
images = page.get_images(full=True)

        for index, img in enumerate(images):

printed img info is like this: (30, 0, 396, 1034, 1, 'DeviceGray', '', 'Im2', 'JBIG2Decode', 0)
How to display this img?

**Comment by nissansz** [comment]
I can judge and delete now with below code.
'''
# Import PyMuPDF
import fitz

# File path you want to extract images from
file = "data.pdf"

# Open the file
pdf_file = fitz.open(file)

# Iterate over PDF pages
for page_index in range(len(pdf_file)):
    # Get the page itself
    page = pdf_file[page_index]

    # Get the image list for the page
    image_list = page.get_images(full=True)

    # Printing the number of images found on this page
    if image_list:
        print(f"[+] Found a total of {len(image_list)} images in page {page_index + 1}")
    else:
        print("[!] No images found on page", page_index + 1)

    # Extract images from the page
    for image_index, img in enumerate(image_list, start=1):
        xref = img[0]
        base_image = pdf_file.extract_image(xref)
        image_bytes = base_image["image"]
        image_ext = base_image["ext"]

        # Save the image to a file
        image_filename = f"page_{page_index + 1}_image_{image_index}.{image_ext}"
        with open(image_filename, "wb") as img_file:
            img_file.write(image_bytes)

# Close the PDF document
pdf_file.close()
''''
